### PR TITLE
add support for structured output

### LINKED ISSUE #3644: Add `json_schema` to `OpenAIChatCompletionClient`
### What feature would you like to be added?

See https://platform.openai.com/docs/guides/structured-outputs/introduction?lang=python

It is currently under beta. 

### Why is this needed?

Useful for structured output.

### PR DESCRIPTION
**Pull request by lordlinus**
## Why are these changes needed?
GPT-4o-2024-08-06 supports structured output with new beta API and this PR will use that API as part of extra_args to return the structured output format. remove stream option since it is not supported

Example usage:
```
    class StructuredOutputPlan(BaseModel):
        initial_goal: str
        steps: List[str]
    result = await self.model_client.create(
        message,
        extra_create_args={"response_format": StructuredOutputPlan},
    )
```

## Related issue number
Resolves #3644 

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by ekzhu** [comment]
Updated the branch to start running CI checks.

File: python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py [hunk 1]
```diff
@@ -401,6 +401,10 @@ async def create(
                     **create_args,
                 )
             )
+        elif "response_format" in create_args:
```
Comment (ekzhu) [inline python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:404]: Instead performing the switching under the hood, can we make it explicit for users by having a `beta` flag in the model client constructor to indicate whether they are using beta features.

File: README.md [hunk 2]
```diff
@@ -287,6 +287,24 @@ Use GitHub [Discussions](https://github.com/microsoft/autogen/discussions) for g
   </a>
 </p>
 
+### How to make use of structured output announced by OpenAI
```
[the remark this answers is not in the document]
Comment (lordlinus) [reply README.md:290]: done added sample notebook with example

File: python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py (hunk 1 above)
Comment (lordlinus) [reply python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py]: done

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekzhu) [reply README.md:290]: Thanks. We can remove this section from README.

File: python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py [hunk 3]
```diff
@@ -526,7 +552,9 @@ async def create_stream(
             )
         else:
             stream_future = asyncio.ensure_future(
-                self._client.chat.completions.create(messages=oai_messages, stream=True, **create_args)
```
Comment (ekzhu) [inline python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:529]: Does the beta client also support stream mode?

Comment (lordlinus) [reply python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:529]: streaming option is available but not as an argument and not sure if this is going to be standard way moving forward
reference: https://github.com/openai/openai-python/blob/main/src/openai/resources/beta/chat/completions.py#L236

should i remove the beta client when streaming is enabled?

File: python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py [hunk 4]
```diff
@@ -393,18 +401,36 @@ async def create(
...
             converted_tools = convert_tools(tools)
-            future = asyncio.ensure_future(
-                self._client.chat.completions.create(
-                    messages=oai_messages,
-                    stream=False,
-                    tools=converted_tools,
-                    **create_args,
+            if self._use_beta_client:
+                future = asyncio.ensure_future(
+                    self._client.beta.chat.completions.parse(
```
Comment (jackgerrits) [inline python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:406]: What other ramifications does `parse` vs `create` have in terms of the API call?

Comment (lordlinus) [reply python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:406]: parse is the recommended approach moving forward based on this cookbook - https://cookbook.openai.com/examples/structured_outputs_intro#using-the-sdk-parse-helper

is there an easy way to test the difference between these 2 approaches?

Comment (jackgerrits) [reply python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:406]: Recommended approach for when using structured outputs right? If you just want completions then is `parse` the call you should be making?

It seems to me that they are completely different APIs?

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lordlinus) [reply README.md]: done

File: python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py (hunk 4 above)
Comment (lordlinus) [reply python/packages/autogen-core/src/autogen_core/components/models/_openai_client.py:406]: structured output for sure, but there is no `create` option for completions in beta SDK
https://github.com/openai/openai-python/blob/8467d41376c2b17eae1c78d56b39767e7807cb6f/src/openai/resources/beta/chat/completions.py#L59

**Pull request closed without merging**